Manage external declarative bases

nickw444 opened the pull request:
Fix for Issue #98

svenstaro commented:
Can you make a test for this can perhaps document it a little?

davidism commented:
Overwriting `self.Model` here isn't correct, as this method will be called after the original Model has been created and could potentially have been used.  Instead, the correct way to implement this is probably to support an arbitrary number of bases.

immunda commented:
@nickw444 Any word on documentation or dealing with this in the way @davidism suggests?

davidism commented:
It might be acceptable to just not overwrite `self.Model`, since references to the external bases are (by definition) external.

Another issue is that this will blindly set `query_class` and `query`, rather than checking if they each exist on the target model first.

Overall the idea is fine, it just needs some more work first.

nickw444 commented:
Hello All, I have been on an extended break, Over the coming days I will correctly implement this patch, write the tests, and make the required modifications for this to be pulled.

immunda commented:
Great, look forward to it! :+1:

nickw444 commented:
Good progress, taking @davidism's advice, I no longer override `self.Model`, since indeed, this should be left for Flask-SQLA's internal model.

 Instead, registering a base will add it to reference in `self.external_bases`. Any requirement to access the base should indeed, be performed by doing it directly, rather than via flask-sqla.

For convenience, I have added a property called `bases` which returns a list, containing the internal `self.Model` and any additional registered bases. 

This pull requests supports an arbitrary number of bases, since all bases, including the internal are available as an iter property (`self.bases`). Operations such as `_execute_for_all_tables` now perform for each base registered, including the internal `self.Model`.

I have added some additional test cases for this pull request, however I have disabled one, as I cannot find a solution to fix this minor problem when using raw SQLA. - When using a relationship (dynamically mapped), the query class is SQLA's own (obviously), however I am unable to re-construct this at runtime to patch it to use Flask-SQLA's own patched BaseQuery with convenience methods (pagination, etc). Only minor for now, but is something worth pointing out. I believe a solution for this is to set the session's default query class, however I did not want to do this as this may have unwanted side affects. 

Hopefully we can get this merged soon, cloning my fork via pip makes me nervous every time :P :+1:

gawry commented:
Is this going to be merged? It would be great to be able to use raw sqlalchemy models with flask

pahaz commented:
+1 same question. Is this going to be merged?

pahaz commented:
Hi, find some problem with this patch:

  `ExtraSQLAlchemy class` - is the same as your patch.



    # https://github.com/omab/python-social-auth/blob/master/examples/flask_example/models/user.py
    from social.apps.flask_app.default.models import init_social
    from social.apps.flask_app.default.models import PSABase
    from flask.ext.sqlalchemy import SQLAlchemy, itervalues, string_types, \
        BaseQuery, _QueryProperty
    from pprint import pprint


    class ExtraSQLAlchemy(SQLAlchemy):
        def __init__(self, app=None, use_native_unicode=True,
                     session_options=None):
            super(ExtraSQLAlchemy, self).__init__(app, use_native_unicode,
                                                  session_options)
            self.external_bases = []

        def get_tables_for_bind(self, bind=None):
            """Returns a list of all tables relevant for a bind."""
            result = []
            for Base in self.bases:
                for table in itervalues(Base.metadata.tables):
                    if table.info.get('bind_key') == bind:
                        result.append(table)
            return result

        def _execute_for_all_tables(self, app, bind, operation, skip_tables=False):
            app = self.get_app(app)

            for Base in self.bases:
                if bind == '__all__':
                    binds = [None] + list(app.config.get('SQLALCHEMY_BINDS') or ())
                elif isinstance(bind, string_types) or bind is None:
                    binds = [bind]
                else:
                    binds = bind

                for bind in binds:
                    extra = {}
                    if not skip_tables:
                        tables = self.get_tables_for_bind(bind)
                        extra['tables'] = tables
                    op = getattr(Base.metadata, operation)
                    op(bind=self.get_engine(app, bind), **extra)

        @property
        def bases(self):
            return [self.Model] + self.external_bases

        def register_base(self, Base):
            """Register an external raw SQLAlchemy declarative base.
            Allows usage of the base with our session management and
            adds convenience query property using BaseQuery by default.
            """
            self.external_bases.append(Base)
            for c in Base._decl_class_registry.values():
                print('e', c)
                if isinstance(c, type):
                    if not hasattr(c, 'query') and not hasattr(c, 'query_class'):
                        c.query_class = BaseQuery

                    if not hasattr(c, 'query'):
                        c.query = _QueryProperty(self)


    app = Flask(__name__)
    app.config.from_object('config')                        

    db = ExtraSQLAlchemy(app)
    init_social(app, db.session)
    db.register_base(PSABase)

    pprint([x.name for x in db.metadata.sorted_tables])
    pprint([x.name for x in PSABase.metadata.sorted_tables])
    
    # 
    # output:
    # 
    # ['user']
    # ['social_auth_nonce',
    #  'social_auth_association',
    #  'social_auth_usersocialauth',
    #  'social_auth_code']

    
    # will work incorrect!
    from flask.ext.migrate import Migrate
    migrate = Migrate(app, db)

As you can see, the `db.metadata.sorted_tables` is different for `PSABase` and `db.Model` classes.
Because of this if I use the `Flask-Migrate` or something witch work with `db.metadata` I will get an unexpected behaviour.

nickw444 commented:
@pahaz, The way this patch is implemented is that it works together with the Flask-Sqlalchemy's builtin declarative base - It does not actually replace the original base. This means you can use a hybrid of both SQLA models and Flask-SQLA models. Additionally, you can register multiple bases.

Hence when needing to access the sorted_tables property, you must do this via each of your declarative bases individually. Alternatively, a property which returns all the bases registered can be found on `db.bases`.

davidism commented:
I think migrations need to be split up per metadata for Alembic anyway.  That should probably be something Flask-Migrate handles, like it handles tying in the one metadata right now.

nickw444 commented:
I agree with @davidism on that point

tony commented:
Yep Houston we have a problem. I need this.

I love flask, but I'm missing the ability to scan and inject the connection data into the objects...

@nickw444 You still on this? I'm going to take a look at how django / django orm does it, then see how this compares.

nickw444 commented:
@tony  I'm still on this, we're just waiting for this to be pulled back. For the time being my fork works flawlessly, I'm using it in about 5 projects

tony commented:
@nickw444 Can you rebase / squash?

I'm going to take a deeper look and give it a QA

tony commented:
@immunda , @davidism : This is working great for me.

@nickw444: Now it's being used on another project.  Thanks for adding tests too.

makmanalp commented:
@tony thanks for the heads up! @nickw444 thanks for the patch! I'll also try running with this flask version and see what happens.

nickw444 commented:
@pahaz I recently needed to use Flask-Migrate with a project with this setup with models. In the alembic `env.py` you need to change the target_metadata variable to point to your actual Base's metadata. By default it's set to `current_app.extensions['migrate'].db.metadata`, which is just the builtin Flask-SQLA metadata.

nickw444 commented:
@tony Are you reviewing on behalf of @mitsuhiko / other devs of this repo? I'll squash now if that's the case. 

Thanks!

nickw444 commented:
@immunda ETA on this being merged/v3? Also am I squashing the changes?

makmanalp commented:
Just an update: have been using this for a while now and no trouble so far.

sterwill commented:
Any update on when this might get merged?  I'd love to be able to remove my models' dependencies on a single module-level "db" (and all the initialization and testing headaches that it brings) and just use plain SQLAlchemy models.

charliewolf commented:
I would also like to know when this is being merged

immunda commented:
I'll be releasing `2.1` this week. After which, I'll merge this into `master` and review what else needs to be included for `3`.

makmanalp commented:
Thank you!

EliFinkelshteyn commented:
+1 on this. I currently have to keep two copies of the same models-- one for SQLAlchemy and one for Flask-SQLAlchemy. Merging this would remove a huge headache. Thanks for doing this @nickw444.

jrhite commented:
@immunda any news on the 2.1 release?

i've also recently run into this problem. i'm using scrapy to scrape data and store models into db via sqlalchemy. later i use flask to serve up the models. scrapy shouldn't/doesn't need to be tied into flask. i want to use the sqlalchemy models in both places.

poked around, but haven't seen any 2.1 yet.

thanks!

jrhite commented:
Another issue I'm seeing is with alembic integration.

I use the formula:

```
db = SQLAlchemy(app)
db.register_base(Base)
```

but this doesn't seem to work with alembic.

For example, I have another manage.py script that uses alembic and basically does:

```
from flask.ext.script import Manager
from flask.ext.migrate import Migrate
from flask.ext.migrate import MigrateCommand

from myapp import app
from myapp import db

from myapp.models import Foo
#from myapp.models import Base


migrate = Migrate(app, db, app.config['MIGRATION_DIR'])
#migrate = Migrate(app, Base, app.config['MIGRATION_DIR'])

manager = Manager(app)

manager.add_command('db', MigrateCommand)

if __name__ == '__main__':
    manager.run()
```

alembic doesn't see any of my models or tables, despite calling `db.register_base(Base)` and then importing `db` in my manage.my script.

However if you look at the code above again and uncomment the `import` and `Migrate(...)` lines that use `Base` directly (instead of `db`), alembic works again.

Before I had the need to have Flask-independent models, I just had all my models extend Flask's `db` object provided by `SQLAlchemy(app)`. Alembic worked fine here. But now that I have Flask-independent models and am trying to use `db.register_base()` alembic doesn't see the models.

davidism commented:
@jrhite bring that up on the Flask-Migrate tracker.  He recently added support for multiple binds, so this shouldn't be too much of a stretch.

nickw444 commented:
@jrhite that's a bit weird. I've been using flask-migrate with most of my projects that are using this patch, and haven't had any trouble.

In my __init__.py:

```
import models
from models import auth
from models import core

from models.core import Job

app.db = SQLAlchemy(app)
app.db.register_base(models.Base)

from flask.ext.migrate import Migrate
migrate = Migrate(app, app.db)
```

Then within your env.py for alembic, you need to change the declarative base that is used:

```
from flask import current_app
import models
from models import Base

config.set_main_option('sqlalchemy.url', current_app.config.get('SQLALCHEMY_DATABASE_URI'))
target_metadata = Base.metadata
```

jrhite commented:
@nickw444 

Interesting, thanks. I think maybe my expectations/assumptions of `db.register_base()` may have been wrong.

My alembic env.py is a bit different from yours. I didn't do anything to it, it's just what was created when running `python manage.py db init`

It looks like:

```
<snip>
# add your model's MetaData object here
# for 'autogenerate' support
# from myapp import mymodel
# target_metadata = mymodel.Base.metadata
from flask import current_app
config.set_main_option('sqlalchemy.url', current_app.config.get('SQLALCHEMY_DATABASE_URI'))
target_metadata = current_app.extensions['migrate'].db.metadata
<snip>
```

so target_metadata is getting its metadata from db.metadata.

I thought that's the way the `db.register_base()` patch was supposed to work. That is to say, once I call `db.register_base()`, then all models are visible through `db.metadata`. Is that not correct?

The reason I ask is because my use case is this...I have scraper code that needs to insert into my DB, but it runs outside the context of flask. So the models the scraper needs I've subclassed based off of `Base` (ie: `declarative_base()`).

On the other hand I have lots of models that do only exist in the context of Flask, so I've subclassed them all off of Flask's `db.Model` (via `db = SQLAlchemy(app)`) 

I guess I was assuming after `db.register_base()`, from Flask's perspective or any Flask-based tool's perspective, there would be no difference between db.metadata and Base.metadata. Indeed I was hoping it would be seamless.

Let me know if I've misunderstood the intentions here.

Thanks!

nickw444 commented:
Hey @jrhite 

I understand why you thought it would _just work_. If you have a look up to the earlier comments in this thread, instead of replacing the built-in flask-sqlalchemy metadata, and thus only allowing 1 declarative base, we decided the patch would best work by allowing registration of additional declarative bases. 

This allows people to continue to use pure SQLA as well as Flask-SQLa side by side. 

As @davidism suggested, you should bring raise this as an issue with flask-migrate/alembic developers, since Flask-SQLA now can have multiple declarative bases, which are accessible via the iter property `bases`. This includes both externally registered bases, as well as the builtin flask-sqlalchemy metadata. 

Hope that clears some things up for you :)

jrhite commented:
@nickw444 

Many thanks for the clarification. Very helpful.

And sure, as soon as this is merged, I'll bring it up the the flask-migrate/alembic guys!

davidism commented:
I'm not sure if this is solvable, but here are a few more rough edges caused by this that I've caught while re-reading the patch.

* If the external models use multiple binds, they must either use Flask-SQLAlchemy's approach to binds (the session looks at `bind_key`) or they must use their own session that has the correct hooks for their routing method.  Using their own session goes against the original reason for this issue, but using the Flask-SQLAlchemy way requires prior knowledge that the models will eventually be used with Flask.
* The `query` property is assigned to external models, but this isn't the only way that models can use the query.  Notably, `Table` and `relationship` would still be using the external (and probably default) query class.  So behavior would be different depending on if the query came from `Model.query` or `instance.dynamic_collection`.

npcole commented:
Is there any documentation for how this can be used? e.g. a minimal app example?

Is it definitely going to be merged in to the development branch?

davidism commented:
@npcole I'm sure it will be documented before release if it's merged.  However, the whole point is that nothing should really change from normal usage, except you add `db.register_base(MyBase)` when setting up the app.

npcole commented:
That sounds really ideal!

I hope it can be merged soon.  It would be good to be able to plan on using this. :)

moorchegue commented:
Hi, comrades! I'm also fighting my way through this with Flask-Alembic (which seems superior than -Migrate). So, is there anything that stops this from being merged? And is there any workarounds one can apply in an app code (without patching FA) to make migrations work for now?

rmalleman commented:
This looks great.  Any chance of it being merged?

jmagnusson commented:
I really wish this could be merged in so I could use all the goodies from `Flask-SQLAlchemy` in my projects.

Does all of @davidism's points still stand? I see that support for a custom metadata instance was added in `2.1`.

tony commented:
@nickw444 can you rebase against master

nickw444 commented:
I'll take a look at this in a few days. Quite busy at the moment :)

nickw444 commented:
Rebased against master. Need to perform some manual testing to verify everything is still working right. Also, i do want to sneak in another patch that makes external bases work with multi-binds (which I've hacked into a project where I needed this).

justanr commented on flask_sqlalchemy/__init__.py:
```diff
@@ -1017,6 +1024,22 @@ def __repr__(self):
+        for c in Base._decl_class_registry.values():
+            if isinstance(c, type):
+                if not hasattr(c, 'query') and not hasattr(c, 'query_class'):
+                    c.query_class = BaseQuery
```
Use `self.Query` instead of just BaseQuery to be consistent if a custom query class is used.

makmanalp commented:
So excited!

gpopovic commented:
Any ETA for this?

nickw444 commented:
Just added some documentation with a minimal example for using this pattern. Super keen for 3.0!! 😃 . Addressing my previous comment regarding multi-binds, everything appears to work without an issue, including alembic, with some minor tweaks to enable multi-binding + modifying the mako template. 

Alembic should work out of the box for users using a single bind. 

FYI: I've been using this fork in production for a bit under a year and a half now.

davidism commented on flask_sqlalchemy/__init__.py:
```diff
@@ -965,20 +971,21 @@ def get_binds(self, app=None):
-            op = getattr(self.Model.metadata, operation)
-            op(bind=self.get_engine(app, bind), **extra)
+        for Base in self.bases:
+            if bind == '__all__':
```
This `if/else` block should move outside the `for` loop because it will be the same for every iteration.

davidism commented on docs/models.rst:
```diff
@@ -143,3 +143,64 @@ Here we configured `Page.tags` to be a list of tags once loaded because we
+Flask-SQLAlchemy.
+
+We declare our model just as you would above, however, using SQLAlchemy 
+constructs, rather than accessing classes via ``db.*``. 
```
Mention `register_base` here, it should link to api docs.

davidism commented:
@nickw444 still need to discuss my three points from https://github.com/mitsuhiko/flask-sqlalchemy/pull/250#issuecomment-149322339

1. can just be documentation "if you're using binds, you'll need to use your own session that knows how to handle them, or use the same `Model.__bind__` system and register the extra engines with `SQLALCHEMY_BINDS`"

2. for each model can check `SomeModel.some_relationship.prop.query_class` and set the custom class if it isn't set.  This only works if an instance of the class hasn't accessed the property yet, otherwise the original query class is cached in some deep undocumented SQLAlchemy place.  That shouldn't be an issue though.

3. Not as concerned with this right now, user could always just pass `declarative_base(metadata=my_metadata)`, but it would be nice to have.

nickw444 commented:
@davidism, sounds good.

2: I can confirm this method works, however, after spending 30 mins playing with this, I can patch it if i know what i'm patching:

```python
if hasattr(c , 'rel_dynamic'):
    c.rel_dynamic.prop.query_class = self.Query
```
(Where I have a relationship named rel_dynamic with lazy='dynamic' on a model).

However, moving this into something generic, it seems any means of touching the attribute via `getattr` caches the previously undefined query class, and thus makes our override attempts useless:

```python
for name in dir(c):
    attr = getattr(c, name)
    if type(attr) == orm.attributes.InstrumentedAttribute:
        if hasattr(attr.prop, 'query_class'):
            attr.prop.query_class = self.Query
```

davidism commented:
Huh, that's really weird.  I tried something similar:

~~~python
for rel in inspect(c).relationships:
    if  rel.lazy == 'dynamic' and rel.query_class is None:
        rel.query_class = self.Query
~~~

@zzzeek is there a way to programatically set the query class for a relationship after the relationship has been defined?  Diving deeper is getting too hard to track what's going on.

---

Here's the test code I'm using.  The second print should be `AppenderMyQuery`, not `AppenderQuery

~~~python
from flask import Flask
from flask_sqlalchemy import SQLAlchemy, BaseQuery
import sqlalchemy as sa
from sqlalchemy import orm
from sqlalchemy.ext.declarative import declarative_base

app = Flask(__name__)
app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False

class MyQuery(BaseQuery):
    def get(self, ident):
        print('MyQuery.get')
        return super(MyQuery, self).get(ident)

db = SQLAlchemy(app, query_class=MyQuery)
Base = declarative_base()

class User(Base):
    __tablename__ = 'user'
    id = sa.Column(sa.Integer, primary_key=True)

class Group(Base):
    __tablename__ = 'group'
    id = sa.Column(sa.Integer, primary_key=True)
    users = orm.relationship(User, 'user_group', lazy='dynamic')

user_group = sa.Table(
    'user_group', Base.metadata,
    sa.Column('user_id', sa.ForeignKey(User.id), primary_key=True),
    sa.Column('group_id', sa.ForeignKey(Group.id), primary_key=True)
)


db.register_base(Base)
print(Group.users.prop.query_class)
print(type(Group().users))
~~~

carbn commented:
Hello everyone and thanks for the patch.

I have a situation where I'd like to use 3rd party vanilla SQLAlchemy models in my flask application with some already existing Flask-SQLAlchemy models, but the catch is that the external classes should be using another database. I added the new database to the ```SQLALCHEMY_BINDS``` and added ```__bind_key__``` to the 3rd party models, but it did not work. After some debugging it looks like all my models are using the default database URI which is not correct. Should this be working or am I just doing something incorrectly?

If I turn this around by binding the existing Flask-SQLAlchemy models explicitly to the correct database (using ```__bind_key__```) and set the ```SQLALCHEMY_DATABASE_URI``` to the 3rd party database it works, but I'm planning on moving more models to the external library so it could only be used as a temporary solution.

davidism commented:
@carbn this isn't merged or released yet.

`__bind_key__` only gets set up correctly on subclasses of `db.Model`, you would need to set `MyModel.table.info['bind_key'] = 'my_bind'` manually.

@nickw444 should `__bind_key__` be recognized and set in the table info when calling `register_base`?

carbn commented:
@davidism thanks! I know this isn't released or merged yet but I'm just eager to test it as it happened to answer an issue I had :). Anyway I got it to work using this:

```
def init_db(app):
    db.init_app(app)
    db.register_base(ExternalBase)

    for cls in ExternalBase._decl_class_registry.values():
        if isinstance(cls, type):
            cls.__table__.info['bind_key'] = 'theotherdatabase'

```

This might actually be a cleaner solution as it does not require any changes to the existing 3rd party models.

davidism commented:
If the other models do not define bind keys, and you want to use our bind system, then you'll have to do that anyway.  I was talking about the case where they do define `__bind_key__`, but don't set `info`.  Now that I think about it, that seems like a pretty rare case.

@nickw444 Perhaps `register_base(MyBase, bind='my_bind')` to set up a bind for the base?

nickw444 commented:
Some good points that have been raised by @carbn, I totally agree that __bind_key__ should just work, as it does with flask-sqlalchemy models. I'll have a play around with this on the weekend.

IIRC I am pretty sure i'm using a similar _hack_ to the one that @carbn posted above in one of my projects:

```python
    for cls in ExternalBase._decl_class_registry.values():
        if isinstance(cls, type):
            cls.__table__.info['bind_key'] = 'theotherdatabase'

```

I'll have a play with this on the weekend and see what I can do.

nickw444 commented:
Just a quick update about something I came across today:

I've started using the app factory pattern for my applications to simplify testing, but this came with an interesting side-affect - It has to do with the Model.query convenience property.

IIRC the convenience `query` property should behave correctly if using the `db.Model` way, since models would be instantiated within a function call, or some scope where they have access to the db for that app.

Unfortunately, we have a nasty side-affect when doing this with raw models - since the models are singleton instances (classes), when we set `Model.query`, it sets it for the current app, and for any future uses. This is because `register_base()` check if query convenience property is set so that it wouldn't be overriding it:

```python
if not hasattr(c, 'query') and not hasattr(c, 'query_class'):
    c.query_class = self.Query
if not hasattr(c, 'query'):
    c.query = _QueryProperty(self)
```

One of the unexpected errors I ran into whilst unit testing:

```
sqlalchemy.exc.InvalidRequestError: Object '<Group at 0x10508dc88>' is already attached to session '2' (this is '4')
```

Which occurred because I loaded an object using Group.query.get_or_404() (and query is bound to an existing app session from a previous testcase), but deleted it using current_app.db.session.delete().

Would it be appropriate to override of the query property when calling `register_base`?

nzjrs commented:
I'd like to use this but am wary of jumping in if the migration / alembic story isn't trustworthy. Is there an Eta for merge and maybe the next flask-sqlalchemy release?

davidism commented:
There is no ETA.  There is also nothing "untrustworthy" about using Flask-SQLAlchemy / Alembic / other base models right now, you're just not going to get automatic integration with Flask-Migrate or some of the helpers from Flask-SQLAlchemy in your external base.

nickw444 commented:
@nzjrs from my experience (running this PR for over a year and a half now in production), I've never had any issues with Alembic. I use flask-migrate the normal way, and just add `import models` (or whatever your models python module is) to the `.mako` template.

Teddy-Schmitz commented:
6 months later, is there any chance of this getting merged in ?

mitsuhiko commented:
@nickw444 sadly this PR will not work with multiple applications safely due the query class set to an application specific one. Same applies to the query attribute.

davidism commented:
@mitsuhiko this PR doesn't change the query class, unless I'm overlooking something.

mitsuhiko commented:
@davidism looking at the code `register_base` sets the model `query_class` and `query` to `self.Query` and `_QueryProperty(self)`.

davidism commented:
But those are extension specific, not app specific. Same with the model, metadata, session class, etc.

mitsuhiko commented:
That however then means that you have to use the model with that instance of the extension only. If that is intended than a security check with an assertion should be added.  People should not assume that this change means they can make one instance of this extension per app then.

davidism commented:
Ah, ok. There's probably a way to make this work.

devyash commented:
Hello, I am facing this exact problem. I auto generated a exisisting mysql schema using sqlacodegen. but the thing is I wish to use flask-Sqlachemy instead pure sqlalchemy. This patch would be really helpfull. I will try to look at it and see if I can contribute.

melchoir55 commented:
If you need it today, you can manually patch it in by overriding one class. I have posted an example to SO here: http://stackoverflow.com/questions/28789063/associate-external-class-model-with-flask-sqlalchemy

The only complication I have had is with flask-migrate. I had to change the generated env script to pull from the metadata on the external base.

devyash commented:
Thank you Issac!

On Thu, Mar 23, 2017 at 9:32 PM, Isaac Martin <notifications@github.com>
wrote:

> If you need it today, you can manually patch it in by overriding one
> class. I have posted an example to SO here: http://stackoverflow.com/
> questions/28789063/associate-external-class-model-with-flask-sqlalchemy
>
> The only complication I have had is with flask-migrate. I had to change
> the generated env script to pull from the metadata on the external base.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/mitsuhiko/flask-sqlalchemy/pull/250#issuecomment-288908724>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMow1uDkovFuv7BzVujX4edylTX0MvGaks5rox1AgaJpZM4DOL8u>
> .
>



-- 
Warm Regards,

Devyash Sanghai
+1-352 284 5228
devyashsanghai@gmail.com
devyashsanghai@ufl.edu
http://devyashsanghai.me/

daviddpark commented:
My team would also like to be able to define SQLAlchemy models in a separate package from the flask app and import the package into both Flask based and applications without a Flask dependency. This seems to be a commonly requested feature.

@mitsuhiko, is there a better approach that you can recommend to enable using SQLAlchemy models that were not defined extending a Flask session's base model?

gtalarico commented:
@daviddpark 

It's possible to use SqlAlchemy outside the flask app as I have been using it.

I used sqlalchemy `declarative_base()` as a Base Class instead of `db.Model` and pass this object as a `model_class` to the Flask-SqlAlchemy constructor.

Maybe we can extend the documentation to include this?

It took me some time to get this to work, 
but I arrived at the solution below after stitching together everything I could find on this topic.
PS: Sample code not tested, there might be some mistakes : )

> models.py - common models to be used by flask app and outside application

```
from sqlalchemy.ext.declarative import declarative_base
Base = declarative_base()

class User(Base)
    pass
```

> flask_app.py - Flask-SqlAlchemy Handles all necessary setup

```
from flask import Flask
from flask_sqlalchemy import SQLAlchemy

from models import User

db = SQLAlchemy(model_class=Base)
app = Flask(__name__)
db.init_app(app)
# 

@app.route('/add')
def add_user():
    user = User()
    db.session.add(user)
    db.session.commit()
```

Using outside of flask requires a few simple steps. I keep those in a dedicated `create_session`
1. Create Engine + Session
2. Add Query Property to Models

> external_application.py
```
from sqlalchemy import create_engine
from sqlalchemy.orm import scoped_session, sessionmaker

from models import Base

engine = create_engine('postgres://url')
session = sessionmaker(bind=engine)
db_session = scoped_session(session)

# Adds Query Property to Models - enables `User.query.query_method()`
Base.query = db_session.query_property()

# Create Tables
Base.metadata.create_all(bind=engine)

user = User()
db_session.add(user)
db_session.commit()
print(User.query.all())
```

reritom commented:
Did anything come from this? Three years of development and discussion and then it was closed?

melchoir55 commented:
> Did anything come from this? Three years of development and discussion and then it was closed?

@gtalarico's comment above is a code example for how to do this today
